From 9d52cffc72397cf297cca79234fa659c235ebe31 Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Thu, 17 Sep 2026 18:16:58 +0800
Subject: [PATCH 3/3] AOSCOS: bfd: prefer "traditional" MIPS input format

The Linux kernel links O32 vDSO with something like:

    ${CROSS_COMPILE}ld -m elf64ltsmip -T vdso-o32.lds *-o32.o

The vdso-o32.lds script sets OUTPUT_FORMAT so it's "not needed" to pass
-melf32ltsmip.  But OUTPUT_FORMAT only specifies the format of the
output file, not the format of the input file.

The default logic to deduce the input format is:

1. try the one corresponding to the emulation
2. try the ones associated with the target triple at build time
3. try all the input formats supported

Here 1 won't work as the emulation is elf64ltsmip and the corresponding
format is elf64-tradlittlemips which cannot work for MIPS O32 objects.
In 2 if binutils was configured with --target=mips64el-linux-gnuabi64,
the associated format list will contain elf32-tradlittlemips before
elf32-littlemips.  So elf32-tradlittlemips will be selected and things
will work normally.

But now we are using /usr/bin/ld -m elf64ltsmip as
mips64el-linux-gnuabi64 in binutils+cross-loongson3, and then all the
associated input formats are in fact associated with the *host* triple.
As 2 fails, the logic falls back to 3 but unfortunately in the list of
all the input formats elf32-littlemips takes precedence over
elf32-tradlittlemips.  As there's no way to distinguish if a relocatable
object file is elf32-littlemips or elf32-tradlittlemips (they only have
a difference for core dump files), the former will be selected but it
mismatches with OUTPUT_FORMAT in the linker script, causing the linker
to error out.

Move all the "traditional" (i.e. SysV) formats before the SGI formats to
mitigate the issue for AOSC OS.  The proper (upstreamable) fix might be
adding a way to specify the "associated formats" at runtime, or guess
them from the emulation instead of relying on the list hard coded during
the build.

Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 bfd/targets.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/bfd/targets.c b/bfd/targets.c
index fcf650410dc..213660cff69 100644
--- a/bfd/targets.c
+++ b/bfd/targets.c
@@ -1168,10 +1168,6 @@ static const bfd_target * const _bfd_target_vector[] =
 	&mips_ecoff_le_vec,
 	&mips_ecoff_bele_vec,
 #ifdef BFD64
-	&mips_elf32_be_vec,
-	&mips_elf32_le_vec,
-	&mips_elf32_n_be_vec,
-	&mips_elf32_n_le_vec,
 	&mips_elf32_ntrad_be_vec,
 	&mips_elf32_ntrad_le_vec,
 	&mips_elf32_ntradfbsd_be_vec,
@@ -1180,14 +1176,18 @@ static const bfd_target * const _bfd_target_vector[] =
 	&mips_elf32_trad_le_vec,
 	&mips_elf32_tradfbsd_be_vec,
 	&mips_elf32_tradfbsd_le_vec,
+	&mips_elf32_be_vec,
+	&mips_elf32_le_vec,
+	&mips_elf32_n_be_vec,
+	&mips_elf32_n_le_vec,
 	&mips_elf32_vxworks_be_vec,
 	&mips_elf32_vxworks_le_vec,
-	&mips_elf64_be_vec,
-	&mips_elf64_le_vec,
 	&mips_elf64_trad_be_vec,
 	&mips_elf64_trad_le_vec,
 	&mips_elf64_tradfbsd_be_vec,
 	&mips_elf64_tradfbsd_le_vec,
+	&mips_elf64_be_vec,
+	&mips_elf64_le_vec,
 #endif
 
 #ifdef BFD64
-- 
2.55.0

